### PR TITLE
Fix #55

### LINKED ISSUE #55: DotNetty.Buffers IndexOutOfBoundsException on ChannelGroup.WriteAndFlushAsync(Unpooled.Empty)
Haven't determined what the underlying cause of the error is yet, but the following block of code throws an `IndexOutOfBoundsException` inside the `Task` returned on `ChannelGroup.WriteAndFlushAsync`

```csharp
public override async Task<bool> Shutdown()
{
    Func<Task, bool> always = task => task.IsCompleted & !task.IsCanceled & !task.IsFaulted;

    var writeStatus = ChannelGroup.WriteAndFlushAsync(Unpooled.Empty).ContinueWith(always); //throws inside Task
    var disconnectStatus = ChannelGroup.DisconnectAsync().ContinueWith(always);
    var closeStatus = ChannelGroup.CloseAsync().ContinueWith(always);
    //_eventListener.Dispose();
    return await Task.WhenAll(writeStatus, disconnectStatus, closeStatus).ContinueWith(tr =>
    {
        if (tr.IsFaulted || tr.IsCanceled) return false;
        return tr.Result.All(x => x);
    });
}
```

### PR DESCRIPTION
**Pull request by nayato**
Motivation:
revisit EmptyByteBuffer port to ensure complete functionality

Modifications:
- ported trivial implementation of ToArray, Duplicate and other methods
- minor cleanup

Result:
EmptyByteBuffer does not throw exceptions when operation can be completed in a trivial way (0 byte length is enough to complete the operation).

### CODE REVIEW

**Comment by nayato** [comment]
@Aaronontheweb PTAL

File: src/DotNetty.Transport/Channels/Groups/DefaultChannelGroup.cs [hunk 1]
```diff
@@ -310,18 +310,19 @@ public Task NewCloseFuture(IChannelMatcher matcher)
 
         static object SafeDuplicate(object message)
         {
-            if (message is IByteBuffer)
+            var buffer = message as IByteBuffer;
+            if (buffer != null)
             {
-                return ((IByteBuffer)message).Duplicate().Retain();
+                return buffer.Duplicate().Retain();
```
Comment (Aaronontheweb) [inline src/DotNetty.Transport/Channels/Groups/DefaultChannelGroup.cs]: IMHO, we should really have a test case to verify that this actually fixes #55

File: src/DotNetty.Transport/Channels/Groups/DefaultChannelGroupCompletionSource.cs [hunk 2]
```diff
@@ -34,7 +34,7 @@ public DefaultChannelGroupCompletionSource(IChannelGroup group, Dictionary<IChan
                 this.futures.Add(pair.Key, pair.Value);
                 pair.Value.ContinueWith(x =>
                 {
-                    bool success = x.IsCompleted && !x.IsFaulted && !x.IsCanceled;
+                    bool success = x.Status == TaskStatus.RanToCompletion;
```
Comment (Aaronontheweb) [inline src/DotNetty.Transport/Channels/Groups/DefaultChannelGroupCompletionSource.cs]: just for my own benefit, this statement and the previous statement are equivalent? "Completed with no errors or cancellation?"

File: src/DotNetty.Transport/Channels/Groups/DefaultChannelGroup.cs (hunk 1 above)
Comment (nayato) [inline src/DotNetty.Transport/Channels/Groups/DefaultChannelGroup.cs]: this change specifically does nothing, it's minor cleanup. Actual change to fix out of range exceptions are in `EmptyByteBuffer`

File: src/DotNetty.Transport/Channels/Groups/DefaultChannelGroupCompletionSource.cs (hunk 2 above)
Comment (nayato) [inline src/DotNetty.Transport/Channels/Groups/DefaultChannelGroupCompletionSource.cs]: correct, only difference is that we're doing a single volatile read with the updated version.

**Comment by mtuchkov** [comment]
It must be OK. In offline discussion with @nayato we agreed that we need a general-purpose test that would cover greater number of cases such as byte buffers of different sizes, including corner cases like 0 bytes, different presets, etc. That can be a separate work item.

**Comment by mtuchkov** [comment]
Please bump the version and update the nuspec as well.
